lightning-invoice/src/utils.rs

```diff
@@ -218,6 +189,95 @@ where
+/// in the invoice.
+///
+/// The filtering ensures that the `RouteHints` returned only include the highest inbound capacity
+/// channel per counterparty node. If any channel with a higher inbound capacity than the given
+/// `min_inbound_capacity_msat` exists, channels to other counterparty nodes with a lower inbound
+/// capacity than `min_inbound_capacity_msat` will be filtered out, even if they are the highest
+/// inbound capacity channel for that specific counterparty node.
+/// If a mix between mix public and private the channels exists, the function returns no RouteHints,
+/// and the sender will need to look at the public channels to find a path instead.
+///
+/// Input:
+/// `channels`: The channels to filter.
+/// `min_inbound_capacity_msat`: Defines the lowest inbound capacity channels must have to not
+/// be filtered out, if any other channel above that amount exists.
+///
+/// Result:
+/// `Vec<RouteHint>`: The filtered `RouteHints`. If empty, the channels is a mix between public and
+/// private.
+fn filter_channels(channels: Vec<ChannelDetails>, min_inbound_capacity_msat: Option<u64>) -> Vec<RouteHint>{
+	let mut filtered_channels: HashMap<PublicKey, (u64, RouteHint)> = HashMap::new();
+	let min_inbound_capacity = min_inbound_capacity_msat.unwrap_or(0);
+
+	let mut is_public_values: HashSet<bool> = HashSet::new();
+	let mut min_capacity_channel_exists = false;
+	let route_hint_from_channel = |channel: &ChannelDetails, short_channel_id: u64, forwarding_info: &CounterpartyForwardingInfo| {
+		RouteHint(vec![RouteHintHop {
+			src_node_id: channel.counterparty.node_id,
+			short_channel_id,
+			fees: RoutingFees {
+				base_msat: forwarding_info.fee_base_msat,
+				proportional_millionths: forwarding_info.fee_proportional_millionths,
+			},
+			cltv_expiry_delta: forwarding_info.cltv_expiry_delta,
+			htlc_minimum_msat: None,
+			htlc_maximum_msat: None,}])
+	};
+	for channel in channels {
+		let short_channel_id = match channel.short_channel_id {
+			Some(id) => id,
+			None => continue,
+		};
+		let forwarding_info = match &channel.counterparty.forwarding_info {
+			Some(info) => info,
+			None => continue,
+		};
+
+		is_public_values.insert(channel.is_public);
+
+		if channel.inbound_capacity_msat >= min_inbound_capacity {
+			min_capacity_channel_exists = true;
+		};
+		match filtered_channels.entry(channel.counterparty.node_id) {
+			hash_map::Entry::Occupied(mut entry) => {
+				let current_max_capacity = entry.get().0;
+				if channel.inbound_capacity_msat < current_max_capacity {
+					continue;
+				}
+				entry.insert((
+					channel.inbound_capacity_msat,
+					route_hint_from_channel(&channel, short_channel_id, forwarding_info),
+				));
+			}
+			hash_map::Entry::Vacant(entry) => {
+				entry.insert((
+					channel.inbound_capacity_msat,
+					route_hint_from_channel(&channel, short_channel_id, forwarding_info),
+				));
+			}
+		}
+	}
+	if is_public_values.contains(&false) && is_public_values.contains(&true) {
```

Comment:
I don't think we actually care about if there are private channels or not here - if there are *any* public channels, we don't need to return any channels at all.

Reply:
Oh, I must have totally misunderstood in which scenario we want to filter the hints for the invoice then, I'm sorry about that! 

Just to clarify based on this feedback: For a node with _only_ public channels (_not just a mix_ of public and private channels), we don't want to include any route hints in the invoice then? Even if that node has multiple channels to the same counterparty node, which otherwise would make the hints filtering signal which of those channels is the highest capacity one/has enough capacity for the invoice amt.

Reply:
If we have public channels, then everyone can see the channel values because we've announced which UTXOs they're associated with, so there's no need to hint anything in the invoices.

Reply:
Thanks for clarifying! I'm still learning about lightning, so sorry if I'm a bit confused about when hints should be included.

I'll update the filtering so that no hints are returned if _any_ channel is public, and I'll also need to update a bunch of the tests to account for that.

Reply:
@TheBlueMatt Isn't `ChannelDetails:: inbound_capacity_msat` actually the channel liquidity balance, which is not publicly known?

Reply:
> Isn't ChannelDetails:: inbound_capacity_msat actually the channel liquidity balance, which is not publicly known?

Yes, I'm not sure your point here, though? Are you suggesting we shouldn't sort based on it because it could reveal which channels of ours have relatively higher balance? I think that's okay, actually, we're not leaking all that much and probing could discover this a bit better, I'd think, anyway.

Reply:
No, and I'm not sure how it got linked here, but because of #1278 - our counterparty will generally not actually care which SCID is present, they'll use whatever has enough balance. Its kinda nice to give them the "best" one, but its really not all *that* important. Only LDK hasn't done #1278, as far as I understand.

Reply:
> I'm wondering, and I think this is what @jkczyz's question is referring to as well: Why doesn't it make sense to include route hints when our node only has public channels, if we're using non public info (the inbound capacity) to generate the hints?

Yeah, that was my thinking.

> No, and I'm not sure how it got linked here, but because of #1278 - our counterparty will generally not actually care which SCID is present, they'll use whatever has enough balance. Its kinda nice to give them the "best" one, but its really not all _that_ important. Only LDK hasn't done #1278, as far as I understand.

Hmm... can't these channels be with different counterparties?

Reply:
> Hmm... can't these channels be with different counterparties?

Yes, that's correct!
 
If a node has 2 public channels with 2 different counterparty nodes, and one of the channels inbound capacity is below the invoice amt, the current code will only include the channel to the other counterparty that's above the invoice amount in the hints.
The test `test_channels_with_lower_inbound_capacity_than_invoice_amt_hints_filtering`, shows that behavior in action :)

Reply:
> Why doesn't it make sense to include route hints when our node only has public channels, if we're using non public info (the inbound capacity) to generate the hints?

Ah, I see your point. Yea, I mean that's not entirely unreasonable, but its also not crazy to just have them retry if they try to use a channel that doesn't have enough capacity. My other concern here is that if a node has some public channels there's a good chance they're a routing node and they have *lots* of public channels. One goal is to cut down the number of channels listed in the invoice, which can prevent QR code usage, and its easier to just include nothing if they're a routing node than it is to try to have some channel count limit behavior.

Reply:
Ah, that makes sense! Ok I'll update the PR asap